src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelLockFileModule.java

```diff
@@ -109,51 +130,113 @@ public void afterCommand() {
-            .setModuleExtensions(combinedExtensionInfos)
-            .build();
-
-    // Write the new value to the file, but only if needed. This is not just a performance
-    // optimization: whenever the lockfile is updated, most Skyframe nodes will be marked as dirty
-    // on the next build, which breaks commands such as `bazel config` that rely on
-    // com.google.devtools.build.skyframe.MemoizingEvaluator#getDoneValues.
-    if (!newLockfile.equals(oldLockfile)) {
-      updateLockfile(workspaceRoot, newLockfile);
+
+    Thread updateLockfile =
+        Thread.startVirtualThread(
+            () -> {
+              var notReproducibleExtensionInfos =
+                  combineModuleExtensions(
+                      oldLockfile.getModuleExtensions(),
+                      newExtensionInfos,
+                      /* hasUsages= */ depGraphValue.getExtensionUsagesTable()::containsRow,
+                      /* reproducible= */ false);
+
+              // Create an updated version of the lockfile, keeping only the extension results from
+              // the old lockfile that are still up-to-date and adding the newly resolved
+              // extension results, as long as any of them are not known to be reproducible.
+              BazelLockFileValue newLockfile =
+                  BazelLockFileValue.builder()
+                      .setRegistryFileHashes(
+                          ImmutableSortedMap.copyOf(moduleResolutionValue.getRegistryFileHashes()))
+                      .setSelectedYankedVersions(moduleResolutionValue.getSelectedYankedVersions())
+                      .setModuleExtensions(notReproducibleExtensionInfos)
+                      .build();
+
+              // Write the new values to the files, but only if needed. This is not just a
+              // performance optimization: whenever the lockfile is updated, most Skyframe nodes
+              // will be marked as dirty on the next build, which breaks commands such as `bazel
+              // config` that rely on
+              // com.google.devtools.build.skyframe.MemoizingEvaluator#getDoneValues.
+              if (!newLockfile.equals(oldLockfile)) {
+                updateLockfile(workspaceRoot, newLockfile);
+              }
+            });
+
+    Thread updatePersistentLockfile =
+        Thread.startVirtualThread(
+            () -> {
+              // Results of reproducible extensions do not need to be stored for reproducibility,
+              // but avoiding reevaluations on server startups helps cold build performance.
+              var reproducibleExtensionInfos =
+                  combineModuleExtensions(
+                      oldLockfile.getModuleExtensions(),
+                      newExtensionInfos,
+                      /* hasUsages= */ depGraphValue.getExtensionUsagesTable()::containsRow,
+                      /* reproducible= */ true);
+
+              var persistentRegistryFileHashes =
```

Comment:
I'm a bit confused about why we even have these. I thought this outputBase lockfile was only to be used for caching reproducible extensions. The reason described in the comment doesn't seem particularly convincing to me -- storing these would only be an improvement for cold-server but hot-outputBase builds, which is a marginal benefit for a rather unrelated addition to the otherwise focused outputBase lockfile.

Reply:
When you switch between branches with new or updated dependencies, even a hot server would benefit. It's relatively uncommon to run `bazel clean --expunge` and the additional implementation complexity seems pretty low, which is why that would be good enough of a reason to add this for me.

Reply:
I dropped this for now. 

What do you think of storing the module version -> hashes map in a subdirectory of the repository cache instead? That way it could be shared across workspaces, similar to how the Go module cache works. I could implement that as a follow-up.

Reply:
hmm. that essentially means we're storing "registry file URL -> contents", since the download cache already stores "hash -> contents". That sounds a tiny bit dangerous to me. Right now if the remote changes for whatever reason, people can fix it by deleting their lockfile. The download cache is much less discoverable.

Reply:
That's true at the moment and it's also why I have dropped this for now. I do think that as the Bazel registry model matures, we'll have to strictly prohibit changes to existing modules and enforce this via attestation (and eventually log transparency techniques, to match Go's security guarantees). At that point we could reconsider this.